### PR TITLE
Improve parallel boundary regions

### PR DESCRIPTION
**Pull request by dschwoerer**
Allow to iterate over different regions separately.

Includes tests.

### CODE REVIEW

**Comment by ZedThree** [comment]
The tests are complaining about some problem in the adams bashforth solver, which I don't think was changed here -- possibly bad merge?

File: src/solver/impls/adams_bashforth/adams_bashforth.cxx [hunk 1]
```diff
@@ -544,6 +544,9 @@ int AdamsBashforthSolver::run() {
     // avoid any additional unrequired work associated with run_rhs.
     run_rhs(simtime);
 
+    // Advance iteration number
+    iteration++;
```
Comment (dschwoerer) [inline src/solver/impls/adams_bashforth/adams_bashforth.cxx:548]:
```suggestion
```

**Comment by dschwoerer** [comment]
Failing tests are due to PETSc 3.18 incompatibility, fixed by #2698

**Comment by dschwoerer** [comment]
https://github.com/boutproject/BOUT-dev/commit/6c0d6b487f19a7b50fa69646a2cf6b565c0bc5ad is missing from master - what is the best way forward? Cherry-pick? Release 5.0.1 and merge?

File: src/mesh/impls/bout/boutmesh.cxx [hunk 2]
```diff
@@ -80,8 +83,10 @@ BoutMesh::~BoutMesh() {
   for (const auto& bndry : boundary) {
     delete bndry;
   }
-  for (const auto& bndry : par_boundary) {
-    delete bndry;
+  if (!par_boundary.empty()) {
+    for (const auto& bndry : par_boundary[0]) {
+      delete bndry;
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/impls/bout/boutmesh.cxx:88]: @ZedThree should we start using gsl::owner? Should we silence those warnings?

File: include/bout/mesh.hxx [hunk 3]
```diff
@@ -486,10 +491,12 @@ public:
   virtual void addBoundary(BoundaryRegion* UNUSED(bndry)) {}
 
   /// Get all the parallel (Y) boundaries on this processor
```
Comment (ZedThree) [inline include/bout/mesh.hxx:493]: Please could you expand this docstring to mention the effect of `type` -- the message on your first commit is a good example

**Review by ZedThree — CHANGES_REQUESTED** [review]

**Comment by dschwoerer** [comment]
> No, better to use unique_ptr instead of raw pointers, and then we don't need the delete at all, but I don't know how big of a change that needs -- might be fairly trivial though

I do think it is non-trivial.
A unique_ptr cannot be used, because each region is in different groups of regions. We could switch to shared_ptr. But I think it is non-trivial and it might be easier to also switch all boundaries at the same time, thus probably out of scope for this PR.

File: include/bout/invert_laplace.hxx [hunk 4]
```diff
@@ -51,6 +51,7 @@ class Laplacian;
 #include "bout/dcomplex.hxx"
 
 class Solver;
+class Datafile;
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply include/bout/invert_laplace.hxx]: Removed

File: tests/integrated/test-fci-boundary/makefile
[the remark this answers is not in the document]
Comment (dschwoerer) [reply tests/integrated/test-fci-boundary/makefile:1]: Good catch, thanks

File: src/mesh/impls/bout/boutmesh.hxx [hunk 5]
```diff
@@ -158,8 +158,8 @@ public:
 
   // Boundary regions
   std::vector<BoundaryRegion*> getBoundaries() override;
-  std::vector<BoundaryRegionPar*> getBoundariesPar() override;
-  void addBoundaryPar(BoundaryRegionPar* bndry) override;
+  std::vector<BoundaryRegionPar*> getBoundariesPar(BoundaryParType type) final;
+  void addBoundaryPar(BoundaryRegionPar* bndry, BoundaryParType type) final;
```
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/impls/bout/boutmesh.hxx]: No particular reason. I changed it to `override`

File: include/bout/mesh.hxx (hunk 3 above)
Comment (dschwoerer) [reply include/bout/mesh.hxx]: I replaced the doc string

**Comment by dschwoerer** [comment]
Thanks for the review @ZedThree, I have updated everything :+1:

File: src/mesh/impls/bout/boutmesh.cxx (hunk 2 above)
[the remark this answers is not in the document]
Comment (dschwoerer) [reply src/mesh/impls/bout/boutmesh.cxx]: I switched to shared_ptr

**Comment by dschwoerer** [comment]
@ZedThree @bendudson is there something I can do to help move this along?
A re-review would be great!

**Review by bendudson — APPROVED** [review]
Thanks, @dschwoerer !

**Pull request merged**